src/main/java/org/opentripplanner/service/realtimevehicles/internal/DefaultRealtimeVehicleService.java

```diff
@@ -43,21 +44,30 @@ public void clearRealtimeVehicles(TripPattern pattern) {
   }
 
   @Override
-  public List<RealtimeVehicle> getRealtimeVehicles(TripPattern pattern) {
+  public List<RealtimeVehicle> getRealtimeVehicles(@Nonnull TripPattern pattern) {
     // the list is made immutable during insertion, so we can safely return them
     return vehicles.getOrDefault(pattern, List.of());
   }
 
   @Nonnull
   @Override
-  public OccupancyStatus getVehicleOccupancyStatus(Trip trip) {
-    TripPattern pattern = transitService.getPatternForTrip(trip);
+  public OccupancyStatus getVehicleOccupancyStatus(@Nonnull Trip trip) {
+    return getOccupancyStatus(trip.getId(), transitService.getPatternForTrip(trip));
+  }
+
+  /**
+   * Get the latest occupancy status for a certain trip. Service contains all the vehicles that
+   * exist in input feeds but doesn't store any historical data. This method is an alternative to
+   * {@link #getVehicleOccupancyStatus(Trip)} and works even when {@link TransitService} is not
+   * provided to the service.
+   */
+  public OccupancyStatus getOccupancyStatus(FeedScopedId tripId, TripPattern pattern) {
```

Comment:
That's a really nice solution to the testability problem.

Reply:
Yes although I had to make the method public since the tests are outside of the package. I don't think we can make as clean solution for how to test the `GtfsRealtimeFuzzyTripMatcher` without transit service though...

Reply:
A public getter method that's not part of the interface is a small price to pay in my view.

Reply:
> You could make `FuzzyTripMatcher` an interface and inject your mock implementation.

I'll bring up this topic in the next dev meeting. I think traditionally we have tried to avoid mocking whenever it's possible.

Reply:
When I said mocking, I didn't mean Mockito but simply implementing a one-method-interface in a test.

Reply:
You'd have to extract the interface though.

Reply:
We decided that this doesn't need to be tested